### ISSUE TITLE
error using Mutect2 vcf

### ISSUE THREAD

**Issue by sraorao** [lima1/PureCN#21]
Hi,
I am trying to use PureCN with a vcf generated by Mutect2, but I get an error as follows when I run filterVcfMuTect2(vcf):
`Error in if (fractionContaminated > 0) { :    missing value where TRUE/FALSE needed`

The vcf was obtained in tumour-only mode with the following command for GATK4 (as recommended [here](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.0.0/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php) but without the `--panel-of-normals` option:
```
  gatk Mutect2 \
   -R ref_fasta.fa \
   -I tumor.bam \
   -tumor tumor_sample_name \
   --germline-resource af-only-gnomad.vcf.gz \
   -O tumor_unmatched_m2_snvs_indels.vcf.gz
```
It turns out PureCN is looking for a geno(vcf) field named FA, but this field doesn't exist in my vcf. Using GATK3 MuTect2 vcf also gives the same error. I feel like I may be missing something, so apologies for posting this as an issue, any help would be appreciated!

**Comment by lima1** [comment]
Thanks for the report. I don't think the FA field is the issue, PureCN *should* standardize the fields in assumes (you could check with PureCN:::.readAndCheckVcf("tumor_unmatched_m2_snvs_indels.vcf.gz", "hg19"), there should be now a FA field).

My guess is that the --germline-resource filters most of the germline sites?  The problem is currently that Mutect2 throws away germline sites and we need them. I emailed the M2 developer and they are open to including flags to keep them. 

Can you post the output of the log-file until the crash? **Ideally using the latest developer version**? This will give you the fraction of variants in the VCF that are known germline sites.  

I never tested it, but if you add "--max-population-af  1" to Mutect2, does it keep all the germline sites in gnomAD?

**Comment by sraorao** [comment]
Thanks for getting back so quickly, Markus, and the helpful tips. Updating to the latest version seems to have fixed the previous error (I was trying `filterVcfMutect2(readVcf(vcf))` directly in R earlier, which was skipping your nifty `.readAndCheckVcf()`), but now there is a new error that there is no DB info field. Which is of course true, as they no longer have a --dbsnp flag in GATK4 Mutect2 (even though the [documentation](https://gatkforums.broadinstitute.org/gatk/discussion/comment/46113#Comment_46113) shows it). 

Log with latest version, using the command line interface (Rscript $PURECN/PureCN.R):
```
INFO [2018-03-11 14:51:57] ------------------------------------------------------------
INFO [2018-03-11 14:51:57] PureCN 1.9.31
INFO [2018-03-11 14:51:57] ------------------------------------------------------------
INFO [2018-03-11 14:51:57] Arguments: -normal.coverage.file EU_Ib/EU_Ib.markeddups_coverage_loess.txt -tumor.coverage.file /projects/wes/snvs_mutect2/EU_It/EU_It.markeddups_coverage_loess.txt -seg.file  -vcf.file gatk4/EU_I_combined_sorted.vcf -normalDB  -genome hg38 -sex ? -args.setMappingBiasVcf NULL -args.segmentation NULL,0.005,NULL -sampleid EU_I -min.ploidy 1 -max.ploidy 6 -max.non.clonal 0.2 -log.ratio.calibration 0.1 -model.homozygous FALSE -error 0.001 -interval.file  -gc.gene.file baits_hg38_gcgene.txt -max.segments 300 -plot.cnv TRUE -DB.info.flag DB -model beta -post.optimize FALSE -log.file ./EU_I.log -args.filterVcf <data> -fun.segmentation <data> -test.num.copy <data> -test.purity <data> -speedup.heuristics <data>
INFO [2018-03-11 14:51:57] Loading coverage files...
INFO [2018-03-11 14:52:21] Mean target coverages: 149X (tumor) 40X (normal).
INFO [2018-03-11 14:52:22] Mean coverages: chrX: 76.81, chrY: 147.73, chr1-22: 139.94.
INFO [2018-03-11 14:52:22] Mean coverages: chrX: 20.45, chrY: 39.87, chr1-22: 37.26.
INFO [2018-03-11 14:52:23] No reptiming column in interval.file.
WARN [2018-03-11 14:52:28] tumor.coverage.file and interval.file do not align.
INFO [2018-03-11 14:52:51] Removing 3879 intervals with missing log.ratio.
INFO [2018-03-11 14:52:51] Removing 1 low/high GC targets.
INFO [2018-03-11 14:52:51] Removing 306 targets with low total coverage in normal (< 150.00 reads).
WARN [2018-03-11 14:52:51] No normalDB provided. Provide one for better results.
INFO [2018-03-11 14:52:51] Removing 60486 low coverage (< 15.0000X) targets.
INFO [2018-03-11 14:52:51] Using 199449 intervals (199447 on-target, 2 off-target).
INFO [2018-03-11 14:52:52] AT/GC dropout: 1.13 (tumor), 1.08 (normal). 
WARN [2018-03-11 14:52:52] High GC-bias in normal or tumor. Is data GC-normalized?
INFO [2018-03-11 14:52:52] Loading VCF...
INFO [2018-03-11 14:52:57] Found 184376 variants in VCF file.
INFO [2018-03-11 14:52:58] Removing 4164 triallelic sites.
FATAL [2018-03-11 14:52:58] vcf.file has no DB info field for dbSNP membership. 

FATAL [2018-03-11 14:52:58]  

FATAL [2018-03-11 14:52:58] This is most likely a user error due to invalid input data or 

FATAL [2018-03-11 14:52:58] parameters (PureCN 1.9.31). 

```
I have now re-tested the GATK3 MuTect2 version with a small sample, using the latest version of PureCN, and have realized that the problem is with the dbsnp file. I will update when I have redone the analysis.

Also, is the warning `tumor.coverage.file and interval.file do not align.` anything to worry about?

To summarise:

- GATK3 MuTect2 vcf was probably a false alarm, small test run seems to work fine, at least until `filterVcfMutect2()`
- GATK4 Mutect2 vcf has a new error - DB field missing; there seems to be no way to do this right now with Mutect2

Thank you for your time!

**Comment by lima1** [comment]
Ok, I can add a function that generates a DB field out of the gnomAD annotation. Would you mind sharing an example VCF that contains a handful of germline calls?

In the meantime, you can use bcftools to add the dbSNP annotation (there is a FAQ for that in the main vignette).

And yes, I would recommend using the command line tools as described in the Quick vignette. They are now fairly mature and provide nearly all the standard functionality of the package.

**Comment by lima1** [comment]
And yes, the warning is weird. Did you use the same interval.file in Coverage.R to generate the coverages? I think this warning should only happen when there are small differences (I should probably stop there instead).

You also have a large GC bias, are you sure you are using the correct baits file? I should probably make PureCN more robust, but things like quality filtering are thrown off with wrong interval files.

**Comment by sraorao** [comment]
I used the same interval file, to generate coverage files. I had to convert my baits from hg19 to hg38 with liftover, which seemed fine on a quick look on the UCSC browser, but I do indeed have to check it more thoroughly. 
![eu_ib markeddups_coverage_loess](https://user-images.githubusercontent.com/12804083/37256534-f1186ff6-2553-11e8-8fb7-fe289b5a3c15.png)
The normal sample did start with a very high GC bias.

gatk4 Mutect2 vcf is attached. This is from Mutect2 (without using FilterMutectVariants), annotated  with bcftools.
[gatk4_dbsnp_annotated_fragment.txt](https://github.com/lima1/PureCN/files/1800648/gatk4_dbsnp_annotated_fragment.txt)

**Comment by lima1** [comment]
1.9.32 should now support the POP_AF info field. I was planning to rewrite the somatic prior probability function for the October Bioconductor release to support population allele frequencies. For now, I don't want to change the API so close to the next release and I just hardcoded a POP_AF > 0.001 cutoff. 

I'm not sure how well filterVcfMutect2 works (this requires running the GATK4 variant filtering tools), but make sure to filter the VCF for artifacts as good as possible. If you have suggestions for improving the defaults in PureCN, let me know.

Can you help me understand where and why the coverage file is different from the interval file? You should get for each line in the interval file coverage information. So are there lines dropped or somehow added? Do you see a pattern? I never seen this bug before.

Yeah, that is a pretty terrible GC dropout. How many normals do you have? For the new coverage normalization, it's actually not that terrible to have a few crappy normals.  That way the normalization learns and removes remaining GC biases after the standard loess normalization.

**Comment by sraorao** [comment]
Thank you. Yes, will experiment with filtering.

I've figured out what caused the coverage file problem. I was still using the older version of PureCN when I got that warning, and for some reason the baits_hg38_gcgene.txt file was sorted in alphanumeric order (chr1, chr10, etc.) even though the original bed file was in the correct (chr1, chr2...) sort order; sorting the GRanges objects within R fixed it. This doesn't happen with the new version (1.9.31). 

Btw, when I run IntervalFile.R from the command line, I get the following error:
```
INFO [2018-03-12 20:56:06] Loading PureCN 1.9.31...
INFO [2018-03-12 20:56:06] Processing /media/srao/hdd0_p1/projects/wes/hglft_genome_74a8_60b060.bed...
Error in path.expand(path) : invalid 'path' argument
Calls: normalizePath -> path.expand
Execution halted
```
But the script works fine, if I use exactly the same arguments inside R (i.e., hardcode all arguments as a list to `opt` object). I have checked again and again that this is not because of a typo, and also use tab completion at the command line, so can't understand why this happens. Anyway, I was able to generate the files, and this is probably not a PureCN issue (I have a similar normalizePath error when running superFreq from the command line).

**Comment by sraorao** [comment]
Thanks for the suggestion about panel of normals, I will give that a try soon. I have 3 normals from the same sequencing run. For now, I was trying a quick analysis with just the matched normal as I wanted to compare results with cnvkit and TitanCNA, but will have to do a proper analysis.

**Comment by lima1** [comment]
No idea about the normalizePath error. What happens when you print out the path?  Like flog.info(">>>%s<<<", opt$infile) before normalizePath?

Normals don't have to be from the same sequencing run. In fact, it's better you have some from multiple batches. The more the better, as long as the capture kit and read lengths are the same.

Hope you have a few more samples of better quality for a meaningful benchmark! PureCN was written for challenging FFPE samples, but dropouts like this are pretty much unusable for allele-specific copy number calling, at least with our assays.

**Comment by sraorao** [comment]
flog.info prints the infile parameter correctly. `ls` of that shows my file, so clearly the path is correct. Weird!

Thanks for the suggestions, I will hopefully have a few more normals in a few weeks. These are all FFPE samples, and using Truseq libraries which, according to at least one paper, seem to have high GC bias.

**Comment by sraorao** [comment]
Btw, I am happy for you to close this, as the original issue has been solved. I will try this on another computer to see if the normalizePath error is reproducible and open a separate issue if so. Thanks again for your time!

**Comment by lima1** [comment]
I could reproduce the Calls: normalizePath -> path.expand bug and it should be fixed now.
